File: lab13_schema_management/src/main/java/wind/WindTurbineDataSupplier.java
package wind;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.function.Supplier;

public class WindTurbineDataSupplier implements Supplier<WindTurbineData> {

    private final Random rand = new Random();
    private final int msgsPerSec;
    List<TurbineProperties> turbines = new ArrayList<>();
    List<String> manufacturers = List.of("Vestas", "Siemens", "Gamesa", "Nordex", "Enercon");
    private int msgCount = 0;

    public WindTurbineDataSupplier(int numTurbines, int msgsPerSec) {
        this.msgsPerSec = msgsPerSec;
        for (int i = 0; i < numTurbines; i++) {
            // Vestas V164-7.0 MW specifications
            double maxPower = 7.0 * 1_000_000; // 7 MW = 7,000,000 Watts
            turbines.add(new TurbineProperties(
                    maxPower,
                    "Turbine" + i,
                    "Windpark" + (i % 5),
                    rand.nextDouble(12),
                    manufacturers.get(rand.nextInt(manufacturers.size()))));
        }
    }

    @Override
    public WindTurbineData get() {
        TurbineProperties turbine = turbines.get(rand.nextInt(turbines.size()));

        // Update wind speed with realistic gradual changes
        double windSpeedChange = (rand.nextDouble() - 0.5) * 2.0; // -1 to +1 m/s change
        double newWindSpeed = Math.max(0, Math.min(25, turbine.currentWindSpeed + windSpeedChange));

        // Calculate power output based on wind speed (simplified power curve)
        double newPowerOutput = calculatePowerOutput(newWindSpeed, turbine.maxPower);

        // Add some realistic noise/variation (±5%)
        double noise = 1.0 + (rand.nextDouble() - 0.5) * 0.1;
        newPowerOutput = Math.max(0, Math.min(turbine.maxPower, newPowerOutput * noise));

        // Update turbine wind speed
        turbine.currentWindSpeed = newWindSpeed;

        WindTurbineData windTurbineData = new WindTurbineData(turbine.id,
                turbine.parkId,
                newPowerOutput
        // TODO: Uncomment this for the next exercise
        // , turbine.manufacturer
        );

        if (msgCount >= msgsPerSec && msgsPerSec != -1) {
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            msgCount = 0;
        }
        msgCount++;
        return windTurbineData;
    }

    private double calculatePowerOutput(double windSpeed, double maxPower) {
        // Simplified power curve for wind turbines
        // Cut-in speed: 3 m/s, Rated speed: 12 m/s, Cut-out speed: 25 m/s
        if (windSpeed < 3.0) {
            return 0.0; // Below cut-in speed
        } else if (windSpeed >= 25.0) {
            return 0.0; // Above cut-out speed (turbine stops for safety)
        } else if (windSpeed >= 12.0) {
            return maxPower; // Rated power (12-25 m/s)
        } else {
            // Cubic relationship between wind speed and power (3-12 m/s)
            double normalizedSpeed = (windSpeed - 3.0) / (12.0 - 3.0);
            return maxPower * Math.pow(normalizedSpeed, 3);
        }
    }

    public static class TurbineProperties {
        public final double maxPower;
        public final String id;
        public final String parkId;
        public double currentWindSpeed;
        public final String manufacturer;

        public TurbineProperties(double maxPower, String id, String parkId, double currentWindSpeed,
                String manufacturer) {
            this.maxPower = maxPower;
            this.id = id;
            this.parkId = parkId;
            this.currentWindSpeed = currentWindSpeed;
            this.manufacturer = manufacturer;
        }
    }
}

File: lab13_schema_management/wind_turbine_data_supplier.py
"""Sample data aligned with the Java WindTurbineDataSupplier (simplified power curve)."""
import random
import time
from typing import Any, Dict, List


class WindTurbineDataSupplier:
    def __init__(self, num_turbines: int, msgs_per_sec: int) -> None:
        self._rand = random.Random()
        self.msgs_per_sec = msgs_per_sec
        self._msg_count = 0
        self._manufacturers = ["Vestas", "Siemens", "Gamesa", "Nordex", "Enercon"]
        self._turbines: List[Dict[str, Any]] = []
        for i in range(num_turbines):
            max_power = 7.0 * 1_000_000
            self._turbines.append(
                {
                    "max_power": max_power,
                    "id": f"Turbine{i}",
                    "park_id": f"Windpark{i % 5}",
                    "wind_speed": self._rand.uniform(0, 12),
                    "manufacturer": self._rand.choice(self._manufacturers),
                }
            )

    def _power_output(self, wind_speed: float, max_power: float) -> float:
        if wind_speed < 3.0:
            return 0.0
        if wind_speed >= 25.0:
            return 0.0
        if wind_speed >= 12.0:
            return max_power
        normalized = (wind_speed - 3.0) / (12.0 - 3.0)
        return max_power * (normalized**3)

    def get(self) -> Dict[str, Any]:
        t = self._turbines[self._rand.randint(0, len(self._turbines) - 1)]
        delta = (self._rand.random() - 0.5) * 2.0
        wind_speed = max(0.0, min(25.0, t["wind_speed"] + delta))
        t["wind_speed"] = wind_speed
        raw_power = self._power_output(wind_speed, t["max_power"])
        noise = 1.0 + (self._rand.random() - 0.5) * 0.1
        power = max(0.0, min(t["max_power"], raw_power * noise))
        return {
            "windTurbineId": t["id"],
            "windParkId": t["park_id"],
            "currentPower": power,
            # TODO (schema evolution): uncomment manufacturer after extending the Avro schema
            # "manufacturer": t["manufacturer"],
        }

    def __iter__(self):
        return self

    def __next__(self) -> Dict[str, Any]:
        value = self.get()
        if self._msg_count >= self.msgs_per_sec and self.msgs_per_sec != -1:
            time.sleep(1)
            self._msg_count = 0
        self._msg_count += 1
        return value
